x-pack/plugins/security_solution/server/lib/detection_engine/signals/threat_mapping/get_signals_map_from_threat_index.test.ts

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ThreatMatchQueryType } from './types';
+
+import { getSignalsQueryMapFromThreatIndex } from './get_signals_map_from_threat_index';
+import { getThreatList } from './get_threat_list';
+import { encodeThreatMatchNamedQuery } from './utils';
+import { MAX_NUMBER_OF_SIGNAL_MATCHES } from './enrich_signal_threat_matches';
+
+import { threatSearchParamsMock } from './get_signals_map_from_threat_index.mock';
+
+jest.mock('./get_threat_list', () => ({ getThreatList: jest.fn() }));
+
+const getThreatListMock = getThreatList as jest.Mock;
+
+export const namedQuery = encodeThreatMatchNamedQuery({
+  id: 'source-1',
+  index: 'source-*',
+  field: 'host.name',
+  value: 'localhost-1',
+  queryType: ThreatMatchQueryType.match,
+});
+
+const termsNamedQuery = encodeThreatMatchNamedQuery({
+  value: 'threat.indicator.domain',
+  field: 'event.domain',
+  queryType: ThreatMatchQueryType.term,
+});
+
+export const threatMock = {
+  _id: 'threat-id-1',
+  _index: 'threats-01',
+  matched_queries: [namedQuery],
+};
+
+const termsThreatMock = {
+  _id: 'threat-id-1',
+  _index: 'threats-01',
+  matched_queries: [termsNamedQuery],
+};
+
+getThreatListMock.mockReturnValue({ hits: { hits: [] } });
+
+describe('getSignalsQueryMapFromThreatIndex', () => {
+  it('should call getThreatList to fetch threats from ES', async () => {
+    getThreatListMock.mockReturnValue({ hits: { hits: [] } });
+
+    await getSignalsQueryMapFromThreatIndex({
+      threatSearchParams: threatSearchParamsMock,
+      eventsCount: 50,
+    });
+
+    expect(getThreatListMock).toHaveBeenCalledTimes(1);
+    expect(getThreatListMock).toHaveBeenCalledWith(threatSearchParamsMock);
+  });
+
+  it('should return empty signals map if getThreatList return empty results', async () => {
+    getThreatListMock.mockReturnValue({ hits: { hits: [] } });
+
+    const signalsQueryMap = await getSignalsQueryMapFromThreatIndex({
+      threatSearchParams: threatSearchParamsMock,
+      eventsCount: 50,
+    });
+
+    expect(signalsQueryMap).toEqual(new Map());
+  });
+
+  it('should return signalsQueryMap for signals if threats search results exhausted', async () => {
+    const namedQuery2 = encodeThreatMatchNamedQuery({
+      id: 'source-2',
+      index: 'source-*',
+      field: 'host.name',
+      value: 'localhost-1',
+      queryType: ThreatMatchQueryType.match,
+    });
+
+    // the third request return empty results
+    getThreatListMock.mockReturnValueOnce({
```

Comment:
I've found tests that mock ES responses are often brittle, either failing when they shouldn't or not failing when they should. It's fine to leave these in, but if we make changes to the implementation and find that these tests require a lot of maintenance it may be easier to remove them or refactor them so they don't rely on mock responses as much, e.g. by extracting the ES calls from the encoding logic and testing the encoding logic with unit tests and ES calls would be covered by integration tests.

Reply:
Thanks for highlighting these concerns.

I share the opinion, it might break too easy if implementation of `getSignalsQueryMapFromThreatIndex` changes.

On the other hand, I think not all the cases might be possible to cover easily by functional tests. For example if we want to cover processing of multiple result pages, we would need to populate test index with thousands of documents. So, it might be easier to mock supposed ES response. Especially, given the fact it unlikely changes in future.

I haven't yet encountered issue with ES mocks you mentioned 
> I've found tests that mock ES responses are often brittle, either failing when they shouldn't or not failing when they should. 
But will pay more attention in future to those kind of tests